src/probnum/kernels/_kernel.py

```diff
@@ -1,135 +1,359 @@
+
+    >>> k.matrix(xs)
+    array([[0.        , 0.        , 0.        , 0.        ],
+           [0.        , 0.33333333, 0.66666667, 1.        ],
+           [0.        , 0.66666667, 1.33333333, 2.        ],
+           [0.        , 1.        , 2.        , 3.        ]])
+
+    Inputs to :meth:`Kernel.__call__` are broadcast according to the "kernel
+    broadcasting" rules detailed in the "Notes" section of the :meth:`Kernel._call__`
+    documentation.
+
+    >>> k(xs[:, None, :], xs[None, :, :])  # same as `.matrix`
+    array([[0.        , 0.        , 0.        , 0.        ],
+           [0.        , 0.33333333, 0.66666667, 1.        ],
+           [0.        , 0.66666667, 1.33333333, 2.        ],
+           [0.        , 1.        , 2.        , 3.        ]])
+
+    A shape of ``1`` along the last axis is broadcast to :attr:`input_dim`.
+
+    >>> xs_d1 = xs[:, [0]]
+    >>> xs_d1.shape
+    (4, 1)
+    >>> xs_d1
+    array([[0.        ],
+           [0.33333333],
+           [0.66666667],
+           [1.        ]])
+    >>> k(xs_d1[:, None, :], xs_d1[None, :, :])  # same as `.matrix`
+    array([[0.        , 0.        , 0.        , 0.        ],
+           [0.        , 0.33333333, 0.66666667, 1.        ],
+           [0.        , 0.66666667, 1.33333333, 2.        ],
+           [0.        , 1.        , 2.        , 3.        ]])
+    >>> k(xs[:, None, :], xs_d1[None, :, :])  # same as `.matrix`
+    array([[0.        , 0.        , 0.        , 0.        ],
+           [0.        , 0.33333333, 0.66666667, 1.        ],
+           [0.        , 0.66666667, 1.33333333, 2.        ],
+           [0.        , 1.        , 2.        , 3.        ]])
+
+    No broadcasting is applied if both inputs have the same shape. For instance, one can
+    efficiently compute just the diagonal of the kernel matrix via
+
+    >>> k(xs, xs)
+    array([0.        , 0.33333333, 1.33333333, 3.        ])
+    >>> k(xs, None)  # x1 = None is an efficient way to set x1 = x0
+    array([0.        , 0.33333333, 1.33333333, 3.        ])
+
+    and the diagonal above the main diagonal of the kernel matrix is retrieved through
+
+    >>> k(xs[:-1, :], xs[1:, :])
+    array([0.        , 0.66666667, 2.        ])
     """
 
-    # pylint: disable="invalid-name"
     def __init__(
         self,
         input_dim: IntArgType,
-        output_dim: IntArgType = 1,
+        shape: ShapeArgType = (),
     ):
-        self._input_dim = np.int_(_utils.as_numpy_scalar(input_dim))
-        self._output_dim = np.int_(_utils.as_numpy_scalar(output_dim))
+        self._input_dim = int(input_dim)
+
+        self._shape = _pn_utils.as_shape(shape)
 
     def __repr__(self) -> str:
         return f"<{self.__class__.__name__}>"
 
-    @abc.abstractmethod
     def __call__(
-        self, x0: _InputType, x1: Optional[_InputType] = None
-    ) -> Union[np.ndarray, np.float_]:
-        """Evaluate the kernel.
+        self,
+        x0: ArrayLike,
+        x1: Optional[ArrayLike],
```

Comment:
```suggestion
        x1: Optional[ArrayLike] = None,
```

Reply:
I believe we agreed that it would be too confusing to have the `x1 == x0` behavior by default here.
Should I change it anyway?